Narrowing to non existing stream gives error message.

krtkmj opened the pull request:
Solves #740

smarx commented:
*Automated message from Dropbox CLA bot*

@krtkmj, it looks like you've already signed the Dropbox CLA. Thanks!

timabbott commented:
This looks pretty good, thanks @krtkmj!  Can you add a quick Casper test for this feature?

krtkmj commented:
@timabbott added casper tests.

timabbott commented:
@krtkmj looks like this is failing some of the Travis CI checks; let me know if you need help understanding them!

timabbott commented:
@umairwaheed do you understand the `check-templates` issue this is getting?

umairwaheed commented:
@timabbott, for some reason Travis is not showing me the console output. I am getting this [error](https://www.dropbox.com/s/jia9ptmrsw02frt/Screen%20Shot%202016-05-11%20at%2012.29.11%20pm.png?dl=0). May be this is some temporary glitch, I'll check again in an hour.

umairwaheed commented:
@timabbott, this PR needs a rebase with `upstream:master` and instead of using `trans` we should use `_()`.

umairwaheed commented on templates/zerver/home.html:
```diff
@@ -42,6 +42,11 @@
         <p>{% trans 'Want to' %} <a href="#" class="empty_feed_join">{% trans 'join it' %}</a>?</p>
       </div>
+      <div id="non_existing_stream_message" class="empty_feed_notice">
+        <h4>{% trans "This stream does not exist. Manage your streams and subscriptions on " %}
```
Instead of `trans` can you please use `_()` function? `upstream:master` now uses Jinja2 and `_()` works in both Django and Jinja2. Thanks.

krtkmj replied:
@umairwaheed Thanks, got it. Working on it

timabbott commented:
This PR looks pretty good!  The only problem is that the narrow.js logic

This doesn't work correctly if you look at a stream that exists but the user has never been subscribed to -- `stream_data` only has data on streams that the user was subscribed to at some point in the past.  A reproducer is to create a new stream as one user in the Zulip development instance, switch to another, and try to narrow to the new stream -- you'll get the "stream does not exist" error.

Ultimately this is I think actually the same bug as we have open in #525 -- that we currently only have present in the Zulip frontend data about streams the user was at one time subscribed to.  

So it seems like it's worth fixing that next.  Here's roughly the approach I'd take to make that work nicely:
1) Move the current `should_list_all_streams` logic into the backend and expose it to the frontend via `page_params`, so we can access it on the backend (we'll want it available in the backend for the following).  I'd add the backend code as a Realm method, e.g. realm.should_list_all_stream().
2) Modify `home` in zerver/views/__init__.py to send the data on public streams that the user is not subscribed to that are needed by `populate_and_fill` in static/js/subs.js as part of page_params, and remove the `/json/streams` call from subs.js to instead poll from `populate_and_fill`.  In this same PR, we may need to add code to update the list of public streams via the Zulip event system (this may involve some tweaks to `apply_events` and code to handle `op=create` and `op=delete` for the "stream" event types in static/js/server_events.js) -- the end goal here is that you should be able to have two browser windows open logged in as different users, both with the streams page open, and in one browser window we create a new stream, and the other should immediately rerender to show that new stream in the list.  And of course this behavior can be off if `should_list_all_streams` is disabled.
3) Then we can tweak this patch to query the now present in the frontend complete data for which streams exist
4) Finally we can add support for sending subscriber counts needed for #525 to the data flow.

Does that make sense @krtkmj?

(as a sidenote, @krtkmj please post a comment on PRs when you've addressed comments and are ready for another review -- GitHub doesn't notify me otherwise).

krtkmj commented:
@timabbott In page_params, there is a list `unsubbed_info` and `subbed_info`. `unsubbed_info` doesn't contain any information of streams that user never subscribed. So, shall we add something like `neversubbed_info` or add never subscribed streams to `unsubbed_info`.

timabbott commented:
Yeah I think `neversubbed_info` is a reasonable name for that.  We may find we want to refactor it all to just one list with field(s) on the streams for what the user's status is, but I would start trying to write it as neversubbed_info; if it's reasonably clean, great; if it gets really messy, it's reasonable to try switching to another approach.

krtkmj commented:
@timabbott 

- Added never subscribed streams info in `neversubbed_info` and exposed it to frontend using page_params
- Never subscribed streams now show corresponding info.

timabbott commented:
@krtkmj it looks like this branch has lint errors (see the Travis CI output), can you fix those?

timabbott commented on zerver/lib/actions.py:
```diff
@@ -2388,16 +2388,22 @@ def gather_subscriptions_helper(user_profile):
     streams_subscribed_map = dict((sub["recipient__type_id"], sub["active"]) for sub in sub_dicts)
     subscriber_map = bulk_get_subscriber_user_ids(streams, user_profile, streams_subscribed_map)
 
+    sub_unsub_stream_ids = []
```
You should use a `set` data structure here for better performance.

neerajwahi commented on zerver/views/streams.py:
```diff
@@ -375,7 +375,6 @@ def get_subscribers_backend(request, user_profile, stream_name=REQ('stream')):
 
     subscribers = get_subscriber_emails(stream, user_profile)
-
     return json_success({'subscribers': subscribers})
```
Remove this hunk. The commit shouldn't be deleting this unrelated line

neerajwahi commented:
@krtkmj looks good overall! Should be pretty much ready to merge once you fix the lint errors.

One interesting case is if a stream is invite-only and the user is not subscribed. Currently this shows 'You are not subscribed to this stream. Want to join it?' but joining it isn't possible without an invite.

Now that we'll have this non-existing stream logic, I think we should either go with the 'Stream does not exist' message in this case, or remove the 'Want to join it?' link. @timabbott thoughts?

timabbott commented:
Yes, agreed; I think probably the "Stream does not exist" error message may be better approach of those.

krtkmj commented:
@timabbott @neerajwahi I guess explicitly mentioning a message "This is invite only stream. And, you aren't invited" because after looking at a message like "Stream does not exist", user might end up trying to create a stream with same name.

krtkmj commented:
@timabbott @neerajwahi Error message for invite only stream updated.

neerajwahi commented:
@krtkmj your test is likely failing because you're not handling the stream creation event in `apply_events`. You need to update `state['never_subscribed']` so that the resulting state from calling `fetch_initial_state_data` + `apply_events` is identical to calling `fetch_initial_state_data` after the event.

krtkmj commented:
@neerajwahi I updated `apply_events` to update `state['never_subscribed']` whenever required. But, tests are still failing. 
What I think is `queries_captured()` which allows a user to capture just the queries execucted. This might be possible that since I updated the code to get never_subscribed streams, number of queries must have reached 14( which in earlier case had be 4).

krtkmj commented:
@neerajwahi One more thing that I noticed is that code for `event['type'] == 'stream'` and `event['op'] == 'create'` code is missing in `apply_events()` is not there. Am I missing something here?

neerajwahi commented:
@krtkmj The `event['op'] == 'create'` logic does not currently exist inside `apply_events()` since previously the creation of a new stream did not change the state. I believe you'll have to add that conditional branch

neerajwahi commented on zerver/lib/actions.py:
```diff
@@ -2388,16 +2388,22 @@ def gather_subscriptions_helper(user_profile):
     for stream in stream_dicts:
         stream_hash[stream["id"]] = stream
 
+    all_streams = list(get_all_realm_streams(user_profile.realm).values())
```
There's no need for this extra query here. You can avoid it by modifying `get_active_streams` above on line 2384 to retrieve all streams (not just streams that the user is subscribed to)

neerajwahi commented on zerver/lib/actions.py:
```diff
@@ -2426,6 +2432,19 @@ def gather_subscriptions_helper(user_profile):
 
+    never_subscribed_stream_ids = [item for item in all_streams_id if item not in sub_unsub_stream_ids]
+    for stream_id in never_subscribed_stream_ids:
+        stream = Stream.objects.get(id=stream_id)
```
You don't need this query as well -- you can look up the stream inside of `stream_dicts` once you make the modification above

neerajwahi commented on zerver/lib/actions.py:
```diff
@@ -2982,6 +3006,9 @@ def get_occupied_streams(realm):
 
     return Stream.objects.filter(id__in=stream_ids, realm=realm, deactivated=False)
 
+def get_all_realm_streams(realm):
```
You can get rid of this once you make the modification to `stream_dicts`

neerajwahi commented on zerver/lib/actions.py:
```diff
@@ -2426,6 +2432,19 @@ def gather_subscriptions_helper(user_profile):
+                       'stream_id': stream.id,
+                       'description': stream.description,
+                       'email_address': encode_email_address_helper(stream.name, stream.email_token)}
+        if subscribers is not None:
```
We shouldn't send back streams that are invite only in `never_subscribed`

neerajwahi commented:
@krtkmj Added some inline comments to the diff.

Basically we can avoid having extra queries here by restructuring the code. We also need to add handlers for `op='create'` and `op='delete'` for the 'stream' event type in `apply_events` (this is why `test_register_events` is failing). Also, we should not include invite only streams in `never_subscribed` and instead treat them as if they don't exist -- we probably don't want to leak even the existence of those streams to someone who wasn't invited.

krtkmj commented:
@neerajwahi Update handlers and optimized query to get never_subscribed streams.

timabbott commented on zerver/lib/actions.py:
```diff
@@ -2569,6 +2576,22 @@ def gather_subscriptions_helper(user_profile):
+        "realm").filter(id__in=never_subscribed_stream_ids)
+    for stream in never_subscribed_streams:
+        if stream.invite_only:
+            continue
```
It'd be cleaner to filter for invite_only streams by adding `, invite_only=False)` to your filter query above.

neerajwahi commented on zerver/lib/actions.py:
```diff
@@ -2524,14 +2524,15 @@ def gather_subscriptions_helper(user_profile):
         stream_hash[stream["id"]] = stream
 
-    all_streams = list(get_all_realm_streams(user_profile.realm).values())
+    all_streams = list(all_active_streams.values())
```
This is still making an extra query. If you remove the `filter(id_in=stream_ids)` from line 2529 then `stream_dicts` will have all active streams and you can use that for `never_subscribed` stream IDs

neerajwahi commented on zerver/lib/actions.py:
```diff
@@ -2576,17 +2577,19 @@ def gather_subscriptions_helper(user_profile):
-        if subscribers is not None:
-            stream_dict['subscribers'] = subscribers
-        never_subscribed.append(stream_dict)
+    never_subscribed_streams = all_active_streams.select_related(
```
This is an unnecessary query. You already queried for the stream data above

neerajwahi commented on zerver/tests/test_subs.py:
```diff
@@ -1469,7 +1469,7 @@ def test_gather_subscriptions(self):
                 continue
             self.assertTrue(len(sub["subscribers"]) == len(users_to_subscribe))
-        self.assert_length(queries, 4, exact=True)
+        self.assert_length(queries, 9, exact=True)
```
In my comment on a previous diff, I meant that you should not have to increase these query limits to add your feature

neerajwahi commented:
@krtkmj Added inline comments. I don't think we need to add any extra queries for this feature

krtkmj commented:
Thanks @neerajwahi. Earlier I thought filtering existing queryset does hit the database again. I was wrong(https://groups.google.com/forum/#!topic/django-users/U3pfYozg4wE). This is great knowledge. 

Now, I can get all necessary information in only 4 queries except to get subscribers of never_subscribed streams I need to make a query. I am looking for a work around. Do you think there is another way without hitting back database for that?

neerajwahi commented on zerver/lib/actions.py:
```diff
@@ -2576,17 +2587,18 @@ def gather_subscriptions_helper(user_profile):
-        if subscribers is not None:
-            stream_dict['subscribers'] = subscribers
-        never_subscribed.append(stream_dict)
+    never_subscribed_streams = [streamx for streamx in all_streams if streamx['id'] in never_subscribed_stream_ids]
```
Why is this named `streamx`?

krtkmj replied:
Naming it stream was raising lint error(stream is already defined above at Line 2561).

neerajwahi commented:
@krtkmj You could modify `subscriber_map` to include all the stream IDs that you care about and avoid an extra query. Look at how the `subscribed` and `unsubscribed` lists are populated

krtkmj commented:
@neerajwahi I can't figure out why this test is failing. Any suggestion?

timabbott commented on zerver/lib/actions.py:
```diff
@@ -1172,6 +1172,17 @@ def get_subscriber_emails(stream, requesting_user=None):
     return [subscription['user_profile__email'] for subscription in subscriptions]
 
+def subscriber_ids_using_stream_dict(user_profile, stream_dict):
+    # type: Iterable[Mapping[str, Any]] -> List[int]
```
This annotation should be
`# type: (UserProfile, Iterable[Mapping[str, Any]]) -> List[int]`
and I think the missing first argument is the cause of the mypy crash in the tests.

neerajwahi replied:
@krtkmj Now that you've updated your diff to use `subscriber_map`, this function is no longer being used and should be removed.

timabbott commented:
Posted a comment that I think explains the cause of the test failure @krtkmj 

@ddfisher FYI this is perhaps more incentive to fix the `--fast-parser` issues around wrong numbers of arguments and/or missing `()` around the argument list

ddfisher commented:
Those are things I'd like to fix soon, but things are already significantly less confusing with the new, correctly-attributed error message.

neerajwahi commented on zerver/lib/actions.py:
```diff
@@ -2568,6 +2579,19 @@ def gather_subscriptions_helper(user_profile):
         else:
             unsubscribed.append(stream_dict)
+    never_subscribed_stream_ids = [item for item in all_streams_id if item not in sub_unsub_stream_ids]
+    never_subscribed_streams = [x for x in all_streams if x['id'] in never_subscribed_stream_ids]
```
Why are you using `x` for the variable name in this list comprehension? `stream` seems more natural here

krtkmj replied:
naming it stream was raising lint error, since stream is already defined in line number 2555

timabbott replied:
Yes, because it's not actually a `Stream` object, it's a dictionary representing a stream.  You could call it `stream_dict` or maybe `ns_stream_dict`.

neerajwahi commented on zerver/lib/actions.py:
```diff
@@ -2737,6 +2763,28 @@ def our_person(p):
+            if event['op'] == 'create':
+                for stream in event['streams']:
+                    stream_name = stream['name']
+                    realm = get_realm('zulip.com')
```
We should not be hard-coding this. What about when we're on a different realm?

krtkmj replied:
This is my bad. Sorry. Fixing it.

neerajwahi commented on zerver/lib/actions.py:
```diff
@@ -2737,6 +2763,28 @@ def our_person(p):
+
+            if event['op'] == 'delete':
+                for stream in event['streams']:
+                    state['stream'] = [x for x in state['stream'] if not x == stream]
```
For consistency with other code in this file, please use `s` instead of `x` as the placeholder variable here (and in the other list comprehensions in this diff)

krtkmj replied:
Shall I remove x from other places as well where I haven't made changes like line 2827.

timabbott replied:
Yeah, that's be reasonable to be do, but it should be done as its own commit (if you rebase to do that as a commit at the start of the branch, I'm happy to merge that cleanup now).

neerajwahi commented on zerver/lib/actions.py:
```diff
@@ -2737,6 +2763,28 @@ def our_person(p):
+                for stream in event['streams']:
+                    stream_name = stream['name']
+                    realm = get_realm('zulip.com')
+                    new_stream, created = create_stream_if_needed(realm, stream_name, stream['invite_only'])
```
Why are you called `create_stream_if_needed` here?

krtkmj replied:
This might be possible that the event is to create a new stream with name of already existing stream. So, if streams with same name already exist this won't create a new stream.

timabbott replied:
This code path should never create a stream -- if you read the flow, a stream event with 'op=create' is generated by the functions that create a stream _after_ they've already created the stream.  The events are to notify browsers about the new stream.

If you need to fetch the stream object, you can use `get_stream`.  But `apply_events` should never contain or call anything that results in a database write (such as create_stream_if_needed).

neerajwahi commented:
@krtkmj Added inline comments

timabbott commented on static/js/narrow.js:
```diff
@@ -525,6 +525,23 @@ function pick_empty_narrow_banner() {
+        if (page_params.neversubbed_info[i].name === stream_name) {
+            never_sub = true;
+        }
+    }
```
I think this could be written more simply with `_.any()`.

neerajwahi commented on static/js/narrow.js:
```diff
@@ -525,6 +525,23 @@ function pick_empty_narrow_banner() {
+
+    // Checks if narrowed to never subscribed stream
+    var never_sub = false;
+    var i;
```
This variable is not being used anymore

neerajwahi commented on static/js/narrow.js:
```diff
@@ -525,6 +525,23 @@ function pick_empty_narrow_banner() {
+    var never_sub = false;
+    var i;
+
+    _.each(page_params.neversubbed_info, function (stream) {
```
I think this would be a lot cleaner if written as a functional style:
```
var never_sub = _.some(page_params.neversubbed_info, function (stream) {
    return stream.name == stream_name;
});

neerajwahi commented on zerver/lib/actions.py:
```diff
@@ -2737,6 +2763,28 @@ def our_person(p):
+                    stream_name = stream['name']
+                    realm = get_realm(user_profile.realm.domain)
+                    new_stream = get_stream(stream_name, realm)
+                    if not stream['invite_only']:
```
Might as well move this check up a few lines, since the preceding variables are not used unless `invite_only === True`

neerajwahi commented on zerver/lib/actions.py:
```diff
@@ -2737,6 +2763,28 @@ def our_person(p):
+
+            if event['op'] == 'delete':
+                for stream in event['streams']:
+                    state['stream'] = [s for s in state['stream'] if not s == stream]
```
I think `s != stream` is clearer than `not s == stream`. Although other list comprehensions use `not x`, they are calling predicate functions instead of using the built-in equality operator.

neerajwahi commented on zerver/lib/actions.py:
```diff
@@ -2515,32 +2515,43 @@ def decode_email_address(email):
+            streams_subscribed_map[stream['id']] = False
+    subscriber_map = bulk_get_subscriber_user_ids(all_streams, user_profile, streams_subscribed_map)
+
+    sub_unsub_stream_ids = []
```
Please make this a `set` for better performance

neerajwahi commented on zerver/lib/actions.py:
```diff
@@ -2515,32 +2515,43 @@ def decode_email_address(email):
     streams_subscribed_map = dict((sub["recipient__type_id"], sub["active"]) for sub in sub_dicts)
-    subscriber_map = bulk_get_subscriber_user_ids(streams, user_profile, streams_subscribed_map)
 
+    for stream in all_streams:
```
This might be simpler as:
```
streams_subscribed_map.update({stream['id']: False for stream in all_streams if stream not in streams})
```

neerajwahi commented:
@krtkmj Thanks for the changes! Added some inline comments. Also your commit title has a typo ('subcibers' instead of 'subscribers')

krtkmj commented:
@neerajwahi Updated this PR.

neerajwahi commented:
@krtkmj Will review your changes now

I just tested this in my VM and noticed an issue. Try the following and see if you can repro:
- Log in as one account and create a public stream, then unsubscribe from that public stream
- Log in as another account and narrow to the public stream you created

The error message offers to let you 'subscribe' to the stream (which doesn't seem to do anything)

krtkmj commented:
@neerajwahi I followed steps you mentioned. One thing that I noticed any stream that has no subscribers is not present in `stream_info` (stream_data.js). And due to this, this stream is not listed in Subscription page(or streams page) for any user. It is possibly considered as inactive stream. And since stream is missing, this results in subs as `undefined` and is possibly creating problem.

@timabbott is there any reason behind this feature or is it a bug?

neerajwahi commented:
@krtkmj Another issue: subscribing to a 'never-subscribed' stream doesn't work.

To reproduce:
- Create a public stream as one user
- Log in as another user, narrow to the new public stream and try to subscribe

I tried this with both empty streams and streams with messages and got the following error in the JS console:
`subs.js:816 Uncaught TypeError: Cannot read property 'subscribed' of undefined`
 
If you look at that line in `subs.js` you'll see that it's checking `sub.subscribed` but `sub` is undefined for this never subscribed stream.

timabbott commented:
@krtkmj a stream that no longer has subscribers is treated as deleted, so that's why it isn't present in `stream_info`.  Is the problem you're thinking about what this page should do with a stream where everyone has been unsubscribed?  It's possible we need to do some tweaks to the model to make this page handle deleted streams correctly.  

I'd like to get at least the backend piece of this merged soon.  @krtkmj can you rebase your branch to have at least two sets of commits:
(1) First, a series of commits that adds the neversubscribed_streams data on the backend, and has appropriate backend tests for this, but doesn't include the UI work for this feature (which is somewhat involved).  My hope is that we can get that finalized and merge it soon, and then you can finish #525 in parallel with finishing the UI work on this.  
(2) Second, a series of commits that implements this feature.

Does that plan seem reasonable @neerajwahi?.

timabbott commented on zerver/lib/actions.py:
```diff
@@ -2537,32 +2537,42 @@ def decode_email_address(email):
-    stream_dicts = get_active_streams(user_profile.realm).select_related(
-        "realm").filter(id__in=stream_ids).values(
-        "id", "name", "invite_only", "realm_id", "realm__domain", "email_token", "description")
+    stream_dicts = [stream for stream in all_streams if stream['id'] in stream_ids]
```
You should switch `stream_ids` to a `set` so that this doesn't take quadratic time

timabbott commented on zerver/lib/actions.py:
```diff
@@ -2761,6 +2789,28 @@ def our_person(p):
+                            'invite_only': new_stream.invite_only,
+                            'name': new_stream.name,
+                            'email_address': encode_email_address_helper(new_stream.name, new_stream.email_token)
+                        }
```
`apply_events` should in theory not do any database queries, because the data it needs should be available in the events themselves.  In the case of stream create events, most of the data in your `stream_data` dict is inside the event (we have a stream.to_dict()), we probably don't need the stream's email address in the never_subscribed list...

neerajwahi commented:
@timabbott That sounds reasonable to me

@krtkmj when you separate out the backend diff into its own commit could you please add some backend tests for `never_subscribed_streams`

krtkmj commented:
@neerajwahi while I was trying to fix this, sub is `undefined` for a stream whenever stream either doesn't exist or user hasn't subscribed to it yet. Now, I am first checking if sub is `undefined` and stream is also present in neversubbed_info, then, subscribe to stream. 

But, one problem that I faced is when a user A creates a stream when another user B is already logged in. Practically, for user B, `page_params.never_subbedinfo` should be updated without refreshing the page. If `never_subbedinfo` is not updated, narrowing through empty narrow feed subscribe button still fails. So, just before I start working how to update `never_subbedinfo` for all users, any suggestion how should I approach it?

neerajwahi commented:
@krtkmj Let's start by separating out the PR into a backend commit and one or more frontend commits as @timabbott suggests above.

Re: your issue above:
If you take a look at https://github.com/zulip/zulip/blob/master/static/js/subs.js#L684, you'll see that `subbed_info` and `unsubbed_info` are both used to populate the `stream_data` data structure and immediately deleted. If there's an update to a stream or subscription, the client is notified via an event and handles the update in this switch statement: https://github.com/zulip/zulip/blob/master/static/js/server_events.js#L56

Your `neversubbed_info` variable is a bit different since it doesn't correspond to a subscription in `stream_data` (and so you'll need to retain it in some fashion). You'll probably have to add a case to `dispatch_event` (in `server_events.js` above) to handle an event of `type=stream` and `op=create` / `op=delete` and update your `neversubbed_info` data structure with the newly created (or newly deleted) stream.

krtkmj commented:
@neerajwahi Rebasing was pretty confusing me. So, I added never subscribed streams backend in another branch. Shall I make a pull request with RFC in header?

neerajwahi commented:
@krtkmj I'd prefer you rebase and keep the commits in this PR, partly to learn how to do it.

Check out this tutorial, specifically the interactive rebase (`git rebase -i`) section: https://www.atlassian.com/git/tutorials/rewriting-history/git-reflog

krtkmj commented:
@neerajwahi Updated this PR.

krtkmj commented:
@neerajwahi Updated this PR.

timabbott commented on zerver/lib/actions.py:
```diff
@@ -2761,6 +2788,27 @@ def our_person(p):
+                            'subscribers':  subscribers,
+                            'invite_only': new_stream.invite_only,
+                            'name': new_stream.name.decode('utf-8')
+                        }
```
I think you can replace the block inside the `invite_only` if statement with:
```
stream_data = copy.deepcopy(stream)
stream_data['subscribers'] = []
state['never_subscribed'].append(stream_data)
```
since structurally, a newly created stream never has any subscribers (even if they are added in the same API call that created the stream, there would be a later event in the series of events that would add the subscribers to the new stream).

timabbott commented on zerver/lib/actions.py:
```diff
@@ -2591,6 +2601,21 @@ def gather_subscriptions_helper(user_profile):
+            stream_dict = {'name': stream['name'],
+                           'invite_only': stream['invite_only'],
+                           'stream_id': stream['id'],
+                           'description': stream['description']}
```
You should write this as `stream_dict = stream.to_dict()`; it does the same thing but duplicates less code :)

krtkmj replied:
stream is already a dictionary. :)

timabbott replied:
Oh, right.  Then you should use `copy.deepcopy` if you want to make a new one, rather than duplicating all the fields.

krtkmj replied:
In this loop, stream is 

    {'name': u'Rome', 'email_token': u'525b13f2dd296cf839ae0182ec3bace1', 
    'realm__domain': u'zulip.com', 'invite_only': False, 'realm_id': 1, 'id': 1, 'description': u''}

Whereas, what we are adding to never_subscribed is 

    {'stream_id': 1, 'description': u'', 'name': u'Rome', 'invite_only': False}. 
So, how can I use `copy.deepcopy` here? Am I missing something?

timabbott replied:
No, I guess there isn't a good cleaner way to do this.  We can stick with the current code.

timabbott commented on zerver/tests/test_subs.py:
```diff
@@ -1556,6 +1556,44 @@ def test_gather_subscriptions(self):
+        invite_only_never_sub = False
+        for stream_dict in never_subscribed:
+            if stream_dict['name'] == "stream_invite_only_1":
+                invite_only_never_sub = True
```
Rather than creating the invite_only_never_sub variable, why not just do 

self.assertFalse(stream_dict['name'] == "stream_invite_only_1")

in the loop?

timabbott commented:
The new backend code is looking really good; I posted a few comments on potential cleanups

krtkmj commented:
@timabbott working on it.

krtkmj commented:
@neerajwahi it's up for review.

krtkmj commented:
@timabbott Shall we merge never subscribed stream backend? Then, I can continue working on #525

timabbott commented:
I took a look, and merged the first commit (the simple variable rename).

However, the second commit has some bugs in how it handles the event updates.  I added two commits on  this branch as part of debugging this: https://github.com/timabbott/zulip/commits/neversubscribed

The first commit adds a test demonstrating the erroneous handling of a deactivated stream the user was not subscribed to, and additionally fixes the code so the test passes.  The second commit adds an additional test that this code causes to not pass, and has some partial work towards fixing the test (it'd be great if you can pick this up and fix the test from here; the issue is that no `peer_add` notice is being sent when another user subscribes to a neversubscribed_stream, and so the subscriber counts aren't actually being live-updated).

timabbott commented:
I'm not 100% sure we actually want to be keeping up-to-date the list of subscribers to `unsubscribed` and `neversubscribed` streams (it is potentially a waste of resources that will be problematic for very large teams; we can do some math to figure that out), so I'm not sure the second commit in my branch is a good direction.  It might be OK.

krtkmj commented:
@timabbott I was working [#d479519](https://github.com/timabbott/zulip/commit/d479519de0d9d84865c7494ca59170e4987c03d0) and realised that in function `do_add_subscription` `peer_add` event is generated but yet tests are failing. Later, just trying somthing out, I replaced `send_event(event, user_ids)` with `send_event(event, active_user_ids(user_profile.realm))` where `user_ids` is empty and later one is list of all users. `test_subscribe_other_user_never_subscribed` test passed. I could see `peer_add` event in `apply_events`. Thoughts?

timabbott commented:
I don't think I understand your question; can you post a link to the relevant code?

krtkmj commented on zerver/lib/actions.py:
```diff
@@ -1415,7 +1415,7 @@ def do_add_subscription(user_profile, stream, no_log=False):
                      subscriptions=[stream.name],
                      user_email=user_profile.email)
-        send_event(event, user_ids)
+        send_event(event, active_user_ids(user_profile.realm))
```
@timabbott In reference to my last [comment](https://github.com/zulip/zulip/pull/749#issuecomment-227593338), I wanted to say that noticed user_ids was `[]`(Hence, user wasn't getting the notice). When I replaced it with `active_user_ids(user_profile.realm)`, `peer_add` event was recieved. I believe it's not the correct way of doing this. Thoughts?
Sorry for that comment

timabbott replied:
yes, agreed, I think we should be using `active_user_ids(user_profile.realm)` here.  The code was correct before we introduced `neversubcribed_streams`, but now that we have that, we need to notify everyone in the realm, not just subscribers, about subscription changes.

krtkmj replied:
Cool. There are few more tests that are failing. I will fix those and update this PR. Sounds okay?

timabbott replied:
Yep!

-Tim Abbott (mobile)
On Jun 25, 2016 13:21, "Kartik Maji" <notifications@github.com> wrote:

> In zerver/lib/actions.py
> <https://github.com/zulip/zulip/pull/749#discussion_r68493400>:
>
> > @@ -1415,7 +1415,7 @@ def do_add_subscription(user_profile, stream, no_log=False):
> >          event = dict(type="subscription", op="peer_add",
> >                       subscriptions=[stream.name],
> >                       user_email=user_profile.email)
> > -        send_event(event, user_ids)
> > +        send_event(event, active_user_ids(user_profile.realm))
>
> Cool. There are few more tests that are failing. I will fix those and
> update this PR. Sounds okay?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/zulip/zulip/pull/749/files/b8588740a9e6b01d69fc546ff5b65f339c186687..c3e93506eddda5c9ce6b08d918a7a85182ea392b#r68493400>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ACnm2pbfb29Pb-doEiDWtxNIXfbS4snHks5qPY2_gaJpZM4IWfUx>
> .
>

krtkmj commented on zerver/tests/test_subs.py:
```diff
@@ -787,7 +787,7 @@ def test_successful_subscriptions_add(self):
             self.helper_check_subs_before_and_after_add(self.streams + add_streams, {},
                 add_streams, self.streams, self.test_email, self.streams + add_streams)
-        self.assert_length(events, 4, True)
+        self.assert_length(events, 6, True)
```
Line no 1374 in `actions.py` results in two more events.

krtkmj commented on zerver/tests/test_subs.py:
```diff
@@ -996,7 +996,7 @@ def test_multi_user_subscription(self):
         self.assert_length(queries, 43)
 
-        self.assert_length(events, 6, exact=True)
+        self.assert_length(events, 8, exact=True)
```
Same as above. Line no 1374 in `actions.py` results in two more events.

krtkmj commented on zerver/tests/test_subs.py:
```diff
@@ -1034,7 +1034,7 @@ def test_multi_user_subscription(self):
         )
 
-        self.assertEqual(len(add_peer_event['users']), 2)
+        self.assertEqual(len(add_peer_event['users']), 14)
```
`peer_add` event is sent to all active realm users.

krtkmj commented:
@timabbott, @neerajwahi : Updated this PR.

neerajwahi commented on zerver/tests/test_subs.py:
```diff
@@ -1560,6 +1566,39 @@ def test_gather_subscriptions(self):
         self.assert_length(queries, 4, exact=True)
 
     @slow(0.15, "common_subscribe_to_streams is slow")
```
You should update this string

neerajwahi commented:
Looks good to me @krtkmj!

@timabbott thoughts? Do we want to add a gate around the new logic so that it doesn't apply for large teams?

timabbott commented:
Sorry for the delay looking at this!  I think the only potential performance concern is `gather_subscriptions_helper` itself (and in particular `bulk_get_subscriber_ids`); the event queue stuff is negligible because people change subscriptions way less than they send messages.  I suspect that might be more expensive than we want it to be for large teams, but I'm not sure what the size cutoff is for that (it might be basically fine).  I can plan to test that and report back on how urgent it is.  

I think probably what we should plan to do for larger teams is just track the number of subscribers, rather than the list of subscribers (and then have the `streams` page query the list of subscribers only when it's needed); that should be inexpensive to handle (and cacheable!).  @krtkmj @neerajwahi can one of you open an issue for doing that as a follow-up?

timabbott commented:
I squashed the backend commits together to try to merge them here: https://github.com/timabbott/zulip/tree/neversubscribed, and noticed a few issues:
* we're sending peer_add events for the users who are being subscribed themselves, which seems like it's going to be duplicate data (see 4a839b003a3d135760cf6c1e25bd6e174b8c4615 for a proposed fix)
* I think we expand the tests to exercise the case of some users having already been subscribed and fully validate that everyone got the correct events here (in particular, I'd like the change I made in 4a839b003a3d135760cf6c1e25bd6e174b8c4615 to have hit a more interesting assert than just the count). 

Otherwise that branch is ready to merge.  @krtkmj can you review my new commit and work on adding the needed test?

krtkmj commented:
@timabbott: Cool, I will work on this and open an issue for the concern your raised above.

krtkmj commented:
@timabbott: I didn't get a clear picture of what new test you want me add here. Can you please explain?

timabbott commented:
Sorry for the slow reply -- what I had in mind is a test similar to `test_never_subscribed_streams` but with some users starting out subscribed and validating that the events produced are correct.

timabbott commented:
I went ahead and merged the main backend commit for this with my tweak, since I think it's good.  You should still add such a test (and finish the frontend for this PR).  Thanks @krtkmj, I'm excited that we have the hard part of this finished!

krtkmj commented:
@timabbott: I have added new test that you were discussing earlier to check whether users are getting correct events. See [0bb99a0](https://github.com/krtkmj/zulip/commit/0bb99a0b27249f8d0a02114c43688429356e9526).

timabbott commented:
https://github.com/krtkmj/zulip/commit/0bb99a0b27249f8d0a02114c43688429356e9526 seems to not be part of this branch?

timabbott commented on static/js/narrow.js:
```diff
@@ -535,6 +551,10 @@ function pick_empty_narrow_banner() {
+    } else if ((first_operator === "stream") && !stream_data.is_subscribed(first_operand)
+                                             && sub.invite_only) {
+        // You are narrowed to a invite only stream to which you aren't subscribed
+        return $("#non_existing_stream_message");
```
I'm not sure we need this check -- is the stream is `invite_only`, the user won't have any data on the stream anyway because of how the backend works, right?  Or am I missing something?

krtkmj replied:
Yeah. user won't be able to see anything of invite_only stream. But, as you mentioned earlier in this discussion, narrowing to a `invite_only` should raise "Stream does not exist" error message.

timabbott commented on static/js/narrow.js:
```diff
@@ -525,6 +525,22 @@ function pick_empty_narrow_banner() {
+    // Checks if narrowed to never subscribed stream
+    var never_sub = _.some(page_params.neversubbed_info, function (stream) {
+        return stream.name === stream_name;
+    });
```
maybe move this `never_sub` calculation inside the `if` statement just below?

krtkmj commented:
@timabbott I updated this PR. For [`0bb99a0`](https://github.com/krtkmj/zulip/commit/0bb99a0b27249f8d0a02114c43688429356e9526) shall I create a new PR or add in this PR?

timabbott commented:
Just add it to this PR.​

krtkmj commented:
@timabbott Updated this PR.

brainwane commented:
@brockwhittaker I was wondering whether you'd take a look at this pull request and give a nonbinding code review?

krtkmj commented:
Rebased this PR.

timabbott commented on static/js/message_list_view.js:
```diff
@@ -509,6 +509,7 @@ MessageListView.prototype = {
         if (stream_name !== undefined) {
             // If user narrows to a stream, doesn't update
             // trailing bookend if user is subscribed.
+            // Update when user is never subscribed(i.e. sub === undefined)
```
(It seems like maybe we should revert the `message_list_view.js` hunk of https://github.com/zulip/zulip/commit/3d77aa49db3492f26accabda95cb84c3ae0069ce)

timabbott commented on zerver/tests/test_subs.py:
```diff
@@ -1399,6 +1399,42 @@ def test_multi_user_subscription(self):
+            for old_user in users_to_subscribe:
+                # Check non new users are in peer_add event recipient list.
+                self.assertIn(old_user, user_dict)
+
```
I merged this commit after a small tweak for clarity; you can drop it from your branch when you next rebase.  Thanks for doing this @krtkmj and sorry for the delay in reviewing this!

timabbott commented on static/js/subs.js:
```diff
@@ -919,10 +919,19 @@ $(function () {
+                            });
+            if (is_never_sub) {
+                ajaxSubscribe(stream_name);
+            }
```
Can you explain the purpose of this change?  It's not clear to me why it wouldn't suffice to instead change just the if statement to `if (sub !== undefined && sub.subscribed) {`, and have that be it?

timabbott replied:
(Also, it seems like maybe we could use the `sub_or_unsub` function and move this improvement into that function?)

zulipbot-test commented:
Hello @krtkmj, a Zulip maintainer reviewed this pull request over a week ago, but you haven't updated your pull request since. Please take a look at the requested changes and update your pull request accordingly.

Thank you for your valuable contributions to Zulip!

zulipbot-test commented:
Hello @krtkmj, a Zulip maintainer reviewed this pull request over a week ago, but you haven't updated your pull request since. Please take a look at the requested changes and update your pull request accordingly.

Thank you for your valuable contributions to Zulip!

timabbott commented:
Closing this, since while it's a useful reference for doing this feature, there's enough merge conflicts that we'll probably want to type it in again before merging.

Thanks for all your work on this @krtkmj!
